File: app/src/main/java/com/kilkari/ui/screens/PaperworkScreen.kt
package com.kilkari.ui.screens

import android.Manifest
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.kilkari.domain.Fmt
import com.kilkari.domain.Paperwork
import com.kilkari.domain.PaperworkStatus
import com.kilkari.domain.PaperworkStep
import com.kilkari.ui.KilkariViewModel
import com.kilkari.ui.components.DetailBar
import com.kilkari.ui.components.GradientCard
import com.kilkari.ui.components.KCard
import com.kilkari.ui.components.KIcons
import com.kilkari.ui.components.KRow
import com.kilkari.ui.components.KSheet
import com.kilkari.ui.components.KSwitch
import com.kilkari.ui.components.SectionLabel
import com.kilkari.ui.nav.NavActions
import com.kilkari.ui.nav.Routes
import com.kilkari.ui.sheets.PaperworkSheet
import com.kilkari.ui.theme.Display
import com.kilkari.ui.theme.KC
import com.kilkari.ui.theme.Sans

/**
 * The four identity documents, in the order they are applied for: birth certificate, then
 * Aadhaar, then passport, then PAN.
 *
 * Only one is ever "next". It sits in the hero card with its due date and a one-tap way to
 * record it; the rail below shows the whole chain, so what has been done and what is still
 * to come are read at a glance. Any step opens to its details, what to carry, and its dates.
 */
@Composable
fun PaperworkScreen(vm: KilkariViewModel, go: NavActions) {
    val steps by vm.paperwork.collectAsStateWithLifecycle()
    val baby by vm.baby.collectAsStateWithLifecycle()
    val reminders by vm.reminders.collectAsStateWithLifecycle()
    val context = LocalContext.current

    /** The key of the step whose sheet is open. */
    var editing by remember { mutableStateOf<String?>(null) }

    // Arriving from a Today row lands straight on that document's sheet.
    val pendingEntry by vm.pendingEntry.collectAsStateWithLifecycle()
    LaunchedEffect(pendingEntry) {
        val entry = pendingEntry ?: return@LaunchedEffect
        if (entry.startsWith("docs:")) {
            editing = entry.removePrefix("docs:")
            vm.consumeEntry()
        }
    }

    val permission = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
        if (!granted) vm.toast("Reminders need notification permission")
    }

    fun ensurePermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return
        val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            android.content.pm.PackageManager.PERMISSION_GRANTED
        if (!granted) permission.launch(Manifest.permission.POST_NOTIFICATIONS)
    }

    val b = baby ?: return
    val active = steps.firstOrNull { it.status == PaperworkStatus.ACTIVE }
    val obtained = steps.count { it.status == PaperworkStatus.OBTAINED }
    val docsReminder = reminders.firstOrNull { it.key == "docs" }

    Box(Modifier.fillMaxSize()) {
        Column(Modifier.fillMaxSize()) {
            DetailBar("Paperwork", go::back)

            Column(
                Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = 16.dp)
                    .padding(top = 10.dp, bottom = 24.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                if (active != null) {
                    NextDocumentCard(
                        step = active,
                        onObtained = { vm.markPaperworkObtained(active) },
                        onDetails = { editing = active.kind.key },
                    )
                } else if (steps.isNotEmpty()) {
                    AllSettledCard(steps, obtained)
                }

                SectionLabel("In this order", Modifier.padding(top = 2.dp))
                KCard {
                    steps.forEachIndexed { i, step ->
                        StepRow(
                            step = step,
                            railAbove = if (i == 0) null else railColour(steps[i - 1]),
                            railBelow = if (i == steps.lastIndex) null else railColour(step),
                            onClick = { editing = step.kind.key },
                        )
                    }
                }

                if (docsReminder != null) {
                    KCard {
                        KRow(
                            title = "Notify me",
                            subtitle = "A week before, the day before, on the day — then weekly while it waits",
                            icon = "notifications_active",
                            iconTint = KC.ClayDeep,
                            iconBg = KC.ClayBg,
                            divider = false,
                            onClick = {
                                if (!docsReminder.enabled) ensurePermission()
                                vm.setReminderEnabled(docsReminder, !docsReminder.enabled)
                            },
                        ) {
                            KSwitch(docsReminder.enabled)
                        }
                    }
                }

                Text(
                    "Each office asks for the document before it, so Kilkari asks for one at a " +
                        "time: the birth certificate ${Paperwork.BIRTH_CERTIFICATE_DAYS} days from " +
                        "${b.name}'s birth, then each of the others a little after the last. The " +
                        "dates are suggestions — open a step to set your own, or to set it aside.",
                    fontFamily = Sans, fontSize = 12.sp, lineHeight = 18.sp, color = KC.Muted,
                )
            }
        }

        val open = steps.firstOrNull { it.kind.key == editing }
        KSheet(open != null, onDismiss = { editing = null }) {
            if (open != null) {
                PaperworkSheet(
                    step = open,
                    dob = b.dob,
                    onSave = { status, settledOn, targetDate, note ->
                        vm.savePaperwork(open, status, settledOn, targetDate, note)
                        editing = null
                    },
                    onScan = {
                        editing = null
                        vm.requestEntry("scan:${open.kind.key}")
                        go.push(Routes.DOCUMENTS)
                    },
                    onOpenDocument = { id ->
                        editing = null
                        vm.selectDocument(id)
                        go.push(Routes.DOCUMENT_DETAIL)
                    },
                )
            }
        }
    }
}

/**
 * The document whose turn it is. Its due date is the headline; recording it is one tap, and
 * everything else — what to carry, the dates, setting it aside — is behind Details.
 */
@Composable
private fun NextDocumentCard(step: PaperworkStep, onObtained: () -> Unit, onDetails: () -> Unit) {
    val days = step.inDays ?: 0
    val due = step.dueDate

    GradientCard(listOf(KC.ClayDeep, KC.Clay, KC.Gold), onClick = onDetails) {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                "NEXT UP · ${step.index + 1} OF ${Paperwork.KINDS.size}",
                fontFamily = Sans, fontWeight = FontWeight.SemiBold,
                fontSize = 12.sp, letterSpacing = 0.5.sp,
                color = Color.White.copy(alpha = 0.85f),
            )
            HeroPill(Fmt.dueText(days), emphasis = days < 0)
        }

        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    step.kind.title,
                    fontFamily = Display, fontWeight = FontWeight.Bold, fontSize = 22.sp,
                    color = Color.White,
                    maxLines = 2, overflow = TextOverflow.Ellipsis,
                )
                Text(
                    if (due != null) "Due ${Fmt.dayAndDate(due)} · ${step.kind.leadText}" else step.kind.leadText,
                    fontFamily = Sans, fontSize = 13.sp, color = Color.White.copy(alpha = 0.9f),
                    maxLines = 2, overflow = TextOverflow.Ellipsis,
                )
            }
            Box(
                Modifier
                    .size(52.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(Color.White.copy(alpha = 0.18f)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(KIcons[step.kind.icon], null, tint = Color.White, modifier = Modifier.size(28.dp))
            }
        }

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) {
            HeroButton("Mark obtained", solid = true, onClick = onObtained)
            HeroButton("Details", solid = false, onClick = onDetails)
        }
    }
}

/** Shown once nothing is left to chase: everything obtained, or what remains set aside. */
@Composable
private fun AllSettledCard(steps: List<PaperworkStep>, obtained: Int) {
    val setAside = steps.filter { it.status == PaperworkStatus.SKIPPED }
    KCard(corner = 16, background = KC.TealBg, border = KC.TealRing) {
        Row(
            Modifier.fillMaxWidth().padding(16.dp),
            horizontalArrangement = Arrangement.spacedBy(14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                Modifier
                    .size(44.dp)
                    .clip(RoundedCornerShape(14.dp))
                    .background(KC.Teal),
                contentAlignment = Alignment.Center,
            ) {
                Icon(KIcons["check"], null, tint = Color.White, modifier = Modifier.size(24.dp))
            }
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(
                    if (setAside.isEmpty()) "All four in hand" else "Nothing waiting",
                    fontFamily = Display, fontWeight = FontWeight.Bold, fontSize = 18.sp, color = KC.TealDeep,
                )
                Text(
                    if (setAside.isEmpty()) {
                        "Birth certificate, Aadhaar, passport and PAN — all obtained."
                    } else {
                        "$obtained of ${steps.size} obtained · " +
                            setAside.joinToString(", ") { it.kind.title } +
                            " set aside. Open one to pick it back up."
                    },
                    fontFamily = Sans, fontSize = 12.sp, lineHeight = 17.sp, color = KC.TealDeep,
                )
            }
        }
    }
}

/**
 * One document on the rail. The node says where it stands; the line running through it
 * turns teal once the chain has passed.
 */
@Composable
private fun StepRow(step: PaperworkStep, railAbove: Color?, railBelow: Color?, onClick: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Column(
            Modifier.width(28.dp).fillMaxHeight(),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Box(Modifier.width(2.dp).height(12.dp).background(railAbove ?: Color.Transparent))
            StepNode(step)
            Box(Modifier.width(2.dp).weight(1f).background(railBelow ?: Color.Transparent))
        }

        Column(Modifier.weight(1f).padding(top = 14.dp, bottom = 14.dp)) {
            Text(
                step.kind.title,
                fontFamily = Sans, fontWeight = FontWeight.SemiBold, fontSize = 14.sp,
                color = when (step.status) {
                    PaperworkStatus.WAITING -> KC.MutedStrong
                    PaperworkStatus.SKIPPED -> KC.Faint
                    else -> KC.Ink
                },
                maxLines = 1, overflow = TextOverflow.Ellipsis,
            )
            Text(
                stepSubtitle(step),
                fontFamily = Sans, fontSize = 12.sp, color = KC.Muted,
                maxLines = 2, overflow = TextOverflow.Ellipsis,
            )
        }

        Box(Modifier.padding(top = 16.dp)) {
            when (step.status) {
                PaperworkStatus.OBTAINED -> Text(
                    "Done", fontFamily = Sans, fontWeight = FontWeight.Bold, fontSize = 12.sp, color = KC.Teal,
                )
                PaperworkStatus.ACTIVE -> Text(
                    Fmt.dueBadge(step.inDays ?: 0),
                    fontFamily = Sans, fontWeight = FontWeight.Bold, fontSize = 12.sp,
                    color = if (step.overdue) KC.Danger else KC.ClayDeep,
                )
                PaperworkStatus.SKIPPED -> Text(
                    "Set aside", fontFamily = Sans, fontSize = 12.sp, color = KC.Faint,
                )
                PaperworkStatus.WAITING -> Icon(
                    KIcons["chevron_right"], null, tint = KC.CoralPaler, modifier = Modifier.size(20.dp),
                )
            }
        }
    }
}

@Composable
private fun StepNode(step: PaperworkStep) {
    val shape = RoundedCornerShape(percent = 50)
    val (background, tint, ring) = when (step.status) {
        PaperworkStatus.OBTAINED -> Triple(KC.Teal, Color.White, KC.Teal)
        PaperworkStatus.ACTIVE -> Triple(KC.Clay, Color.White, KC.ClayBg2)
        PaperworkStatus.SKIPPED -> Triple(KC.StoneBg, KC.StoneMid, KC.StoneBg)
        PaperworkStatus.WAITING -> Triple(KC.Surface, KC.Faint, KC.Track)
    }
    Box(
        Modifier
            .size(28.dp)
            .clip(shape)
            .background(background)
            .border(2.dp, ring, shape),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            KIcons[if (step.status == PaperworkStatus.OBTAINED) "check" else step.kind.icon],
            null, tint = tint, modifier = Modifier.size(15.dp),
        )
    }
}

private fun stepSubtitle(step: PaperworkStep): String = when (step.status) {
    PaperworkStatus.OBTAINED -> listOfNotNull(
        step.settledOn?.let { "Obtained ${Fmt.date(it)}" },
        step.documentTitle?.let { "scan filed" },
    ).joinToString(" · ")
    PaperworkStatus.ACTIVE -> step.dueDate?.let { "Due ${Fmt.dayAndDate(it)} · ${step.kind.leadText}" }
        ?: step.kind.leadText
    PaperworkStatus.WAITING -> step.dueDate?.let { "Your date: ${Fmt.date(it)}" }
        ?: step.previous?.let { "After the ${it.title.lowercase()}" }
        ?: step.kind.leadText
    PaperworkStatus.SKIPPED -> "Set aside · tap to pick it back up"
}

/** The rail turns teal below a step the chain has moved past. */
private fun railColour(step: PaperworkStep): Color = when (step.status) {
    PaperworkStatus.OBTAINED, PaperworkStatus.SKIPPED -> KC.TealLight
    else -> KC.Track
}

@Composable
private fun HeroPill(text: String, emphasis: Boolean) {
    Box(
        Modifier
            .clip(RoundedCornerShape(999.dp))
            .background(if (emphasis) Color.White else Color.White.copy(alpha = 0.18f))
            .padding(horizontal = 9.dp, vertical = 3.dp),
    ) {
        Text(
            text, fontFamily = Sans, fontSize = 12.sp,
            fontWeight = if (emphasis) FontWeight.Bold else FontWeight.Normal,
            color = if (emphasis) KC.Danger else Color.White,
        )
    }
}

@Composable
private fun HeroButton(label: String, solid: Boolean, onClick: () -> Unit) {
    Box(
        Modifier
            .clip(RoundedCornerShape(999.dp))
            .background(if (solid) Color.White else Color.White.copy(alpha = 0.16f))
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp, vertical = 8.dp),
    ) {
        Text(
            label, fontFamily = Sans, fontWeight = FontWeight.Bold, fontSize = 13.sp,
            color = if (solid) KC.ClayDeep else Color.White,
        )
    }
}

File: ios/Kilkari/Screens/PaperworkScreen.swift
import SwiftUI
import SwiftData
import KilkariCore

/// One document's state, kept only when the parent has done something about it.
@Model
final class PaperworkRecord {
    var key: String
    /// The day it was obtained. Nil while it is only set aside.
    var obtainedOn: Date?
    var skipped: Bool

    init(key: String, obtainedOn: Date? = nil, skipped: Bool = false) {
        self.key = key
        self.obtainedOn = obtainedOn
        self.skipped = skipped
    }
}

/// The four documents an Indian child needs, in the order the offices insist on.
///
/// A chain, not a checklist: Aadhaar wants the birth certificate, the passport wants Aadhaar,
/// the PAN wants both. So the screen leads with the one whose turn it is and nothing else,
/// and the rest sit below it as an order to expect rather than four things to do now.
struct PaperworkScreen: View {
    let baby: Baby

    @Environment(\.modelContext) private var context
    @Environment(\.accent) private var accent
    @Query private var records: [PaperworkRecord]

    @State private var editing: PaperworkStep?

    private var steps: [PaperworkStep] {
        var settled: [String: Date] = [:]
        var skipped: Set<String> = []
        for record in records {
            if let on = record.obtainedOn { settled[record.key] = on }
            if record.skipped { skipped.insert(record.key) }
        }
        return Paperwork.chain(birth: baby.dob, settled: settled, skipped: skipped)
    }

    private var active: PaperworkStep? {
        steps.first { $0.status == .active || $0.status == .overdue }
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                if let active {
                    nextUp(active)
                } else {
                    allSettled
                }

                SectionLabel("In this order").padding(.top, 2)
                KCard {
                    VStack(spacing: 0) {
                        ForEach(Array(steps.enumerated()), id: \.element.id) { i, step in
                            stepRow(
                                step,
                                railAbove: i == 0 ? nil : railColour(steps[i - 1]),
                                railBelow: i == steps.count - 1 ? nil : railColour(step)
                            )
                        }
                    }
                }

                Text("""
                Each office asks for the document before it, so Kilkari asks for one at a time: \
                the birth certificate 45 days from \(baby.name)'s birth, then each of the others \
                a little after the last. The dates are suggestions — open a step to record your \
                own, or to set it aside.
                """)
                    .font(KFont.sans(12)).foregroundStyle(KC.muted)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.horizontal, 4)
            }
            .padding(.horizontal, 16).padding(.top, 8).padding(.bottom, 110)
        }
        .sheet(item: $editing) { step in
            PaperworkStepSheet(
                step: step,
                onObtained: { on in settle(step.kind.key, obtainedOn: on); editing = nil },
                onSkip: { settle(step.kind.key, obtainedOn: nil); editing = nil },
                onUndo: { clear(step.kind.key); editing = nil }
            )
            .presentationDetents([.medium, .large])
            .environment(\.accent, accent)
        }
    }

    /// The document whose turn it is. Its date is the headline, recording it is one tap, and
    /// everything else is behind the card.
    private func nextUp(_ step: PaperworkStep) -> some View {
        let index = (steps.firstIndex { $0.id == step.id } ?? 0) + 1
        return GradientCard(colors: step.status == .overdue
                            ? [KC.danger, KC.clayDeep]
                            : [KC.clayDeep, KC.clay]) {
            VStack(alignment: .leading, spacing: 10) {
                HStack {
                    Text("NEXT UP · \(index) OF \(steps.count)")
                        .font(KFont.sans(12, .semibold)).tracking(0.5)
                        .foregroundStyle(.white.opacity(0.85))
                    Spacer()
                    Text(badge(step))
                        .font(KFont.sans(12)).foregroundStyle(.white)
                        .padding(.horizontal, 9).padding(.vertical, 3)
                        .background(.white.opacity(0.18)).clipShape(Capsule())
                }
                Text(step.kind.title)
                    .font(KFont.display(22, .bold)).foregroundStyle(.white)
                Text(step.kind.why)
                    .font(KFont.sans(13)).foregroundStyle(.white.opacity(0.9))
                    .fixedSize(horizontal: false, vertical: true)

                VStack(alignment: .leading, spacing: 5) {
                    ForEach(step.kind.needs, id: \.self) { need in
                        HStack(alignment: .top, spacing: 7) {
                            Circle().fill(.white.opacity(0.7))
                                .frame(width: 4, height: 4).padding(.top, 6)
                            Text(need)
                                .font(KFont.sans(12)).foregroundStyle(.white.opacity(0.9))
                                .fixedSize(horizontal: false, vertical: true)
                        }
                    }
                }

                HStack(spacing: 8) {
                    Button {
                        settle(step.kind.key, obtainedOn: .now)
                    } label: {
                        Text("Got it")
                            .font(KFont.sans(13, .bold))
                            .foregroundStyle(step.status == .overdue ? KC.danger : KC.clayDeep)
                            .padding(.horizontal, 16).frame(height: 38)
                            .background(.white).clipShape(Capsule())
                    }
                    Button {
                        editing = step
                    } label: {
                        Text("Details")
                            .font(KFont.sans(13, .semibold)).foregroundStyle(.white)
                            .padding(.horizontal, 14).frame(height: 38)
                            .background(.white.opacity(0.18)).clipShape(Capsule())
                    }
                }
                .buttonStyle(.plain)
                .padding(.top, 2)
            }
        }
    }

    private var allSettled: some View {
        let done = steps.filter { $0.status == .obtained }.count
        return GradientCard(colors: [KC.leafDeep, KC.leaf]) {
            VStack(alignment: .leading, spacing: 6) {
                Text("ALL SETTLED")
                    .font(KFont.sans(12, .semibold)).tracking(0.5)
                    .foregroundStyle(.white.opacity(0.85))
                Text(done == steps.count ? "All four in hand" : "\(done) in hand, the rest set aside")
                    .font(KFont.display(20, .bold)).foregroundStyle(.white)
                Text("Nothing is waiting on you.")
                    .font(KFont.sans(13)).foregroundStyle(.white.opacity(0.9))
            }
        }
    }

    /// One step in the order, with the rail that ties it to the ones either side.
    ///
    /// The rail is the whole point of the list: these are not four independent errands, and a
    /// plain list of rows would say they were.
    private func stepRow(_ step: PaperworkStep, railAbove: Color?, railBelow: Color?) -> some View {
        Button {
            editing = step
        } label: {
            HStack(alignment: .top, spacing: 12) {
                VStack(spacing: 0) {
                    Rectangle().fill(railAbove ?? .clear).frame(width: 2, height: 10)
                    ZStack {
                        Circle().fill(dotBackground(step)).frame(width: 26, height: 26)
                        Image(systemName: dotSymbol(step))
                            .font(.system(size: 12, weight: .bold))
                            .foregroundStyle(dotTint(step))
                    }
                    Rectangle().fill(railBelow ?? .clear).frame(width: 2).frame(maxHeight: .infinity)
                }
                .frame(width: 26)

                VStack(alignment: .leading, spacing: 2) {
                    Text(step.kind.title)
                        .font(KFont.sans(15, .semibold))
                        .foregroundStyle(step.status == .obtained || step.status == .skipped
                                         ? KC.faint : KC.ink)
                    Text(subtitle(step))
                        .font(KFont.sans(12)).foregroundStyle(KC.muted)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .padding(.top, 8)

                Spacer(minLength: 0)

                Text(badge(step))
                    .font(KFont.sans(11, .semibold))
                    .foregroundStyle(dotTint(step))
                    .padding(.horizontal, 8).padding(.vertical, 3)
                    .background(dotBackground(step)).clipShape(Capsule())
                    .padding(.top, 10)
            }
            .padding(.horizontal, 14)
            .padding(.bottom, 12)
            .frame(minHeight: 62)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private func subtitle(_ step: PaperworkStep) -> String {
        switch step.status {
        case .obtained: return "In hand"
        case .skipped: return "Set aside"
        case .waiting: return step.kind.leadText
        case .active, .overdue: return step.dueOn.map { "By \(Fmt.date($0))" } ?? step.kind.leadText
        }
    }

    private func badge(_ step: PaperworkStep) -> String {
        switch step.status {
        case .obtained: return "Done"
        case .skipped: return "Skipped"
        case .waiting: return "Waits"
        case .active:
            let days = step.inDays ?? 0
            return days == 0 ? "today" : "\(days) \(Fmt.plural(days, "day"))"
        case .overdue:
            let late = -(step.inDays ?? 0)
            return "\(late) over"
        }
    }

    private func dotSymbol(_ step: PaperworkStep) -> String {
        switch step.status {
        case .obtained: return "checkmark"
        case .skipped: return "minus"
        case .overdue: return "exclamationmark"
        case .active: return "arrow.right"
        case .waiting: return "clock"
        }
    }

    private func dotTint(_ step: PaperworkStep) -> Color {
        switch step.status {
        case .obtained: return KC.leafDeep
        case .overdue: return KC.danger
        case .active: return KC.clayDeep
        case .skipped, .waiting: return KC.muted
        }
    }

    private func dotBackground(_ step: PaperworkStep) -> Color {
        switch step.status {
        case .obtained: return KC.leafBg
        case .overdue: return KC.dangerBg
        case .active: return KC.clayBg
        case .skipped, .waiting: return KC.stoneBg
        }
    }

    /// The rail is coloured by what has been settled, so the chain reads as a progress line.
    private func railColour(_ step: PaperworkStep) -> Color {
        switch step.status {
        case .obtained: return KC.leafRing
        case .skipped: return KC.track
        default: return KC.border
        }
    }

    private func settle(_ key: String, obtainedOn: Date?) {
        clear(key)
        context.insert(PaperworkRecord(key: key, obtainedOn: obtainedOn, skipped: obtainedOn == nil))
    }

    private func clear(_ key: String) {
        records.filter { $0.key == key }.forEach(context.delete)
    }
}

/// One step, opened: what it takes, and the three things you can do about it.
struct PaperworkStepSheet: View {
    let step: PaperworkStep
    let onObtained: (Date) -> Void
    let onSkip: () -> Void
    let onUndo: () -> Void

    @Environment(\.dismiss) private var dismiss
    @Environment(\.accent) private var accent

    @State private var on = Date.now

    private var settled: Bool { step.status == .obtained || step.status == .skipped }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    Text(step.kind.why)
                        .font(KFont.sans(14)).foregroundStyle(KC.mutedStrong)
                        .fixedSize(horizontal: false, vertical: true)

                    VStack(alignment: .leading, spacing: 8) {
                        Text("What you need")
                            .font(KFont.sans(13, .semibold)).foregroundStyle(KC.mutedStrong)
                        ForEach(step.kind.needs, id: \.self) { need in
                            HStack(alignment: .top, spacing: 8) {
                                Circle().fill(accent.ring).frame(width: 5, height: 5).padding(.top, 6)
                                Text(need).font(KFont.sans(13)).foregroundStyle(KC.ink)
                                    .fixedSize(horizontal: false, vertical: true)
                            }
                        }
                    }

                    if settled {
                        PrimaryButton(label: "Undo") { onUndo() }
                    } else {
                        // The date is asked for rather than assumed: paperwork is usually
                        // recorded days after the queue it was collected in.
                        DatePicker("Got it on", selection: $on, in: ...Date.now,
                                   displayedComponents: .date)
                            .font(KFont.sans(14, .semibold)).tint(accent.main)

                        PrimaryButton(label: "Record it") { onObtained(on) }

                        Button("Not doing this one") { onSkip() }
                            .font(KFont.sans(14, .semibold))
                            .foregroundStyle(KC.muted)
                            .frame(maxWidth: .infinity)
                            .padding(.top, 2)
                    }

                    Text(step.kind.leadText)
                        .font(KFont.sans(11)).foregroundStyle(KC.faint)
                        .frame(maxWidth: .infinity, alignment: .center)
                }
                .padding(20)
            }
            .background(KC.surface)
            .navigationTitle(step.kind.title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Close") { dismiss() }.tint(accent.deep)
                }
            }
        }
    }
}
